core/src/main/java/io/grpc/RoundRobinLoadBalancerFactory.java

```diff
@@ -0,0 +1,179 @@
+ * addresses from the {@link NameResolver}.  The sub-lists received from the name resolver
+ * are considered to be an {@link EquivalentAddressGroup} and each of these sub-lists is
+ * what is then balanced across.
+ */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/1771")
+public final class RoundRobinLoadBalancerFactory extends LoadBalancer.Factory {
+
+  private static final RoundRobinLoadBalancerFactory instance = new RoundRobinLoadBalancerFactory();
+
+  private RoundRobinLoadBalancerFactory() {
+  }
+
+  public static RoundRobinLoadBalancerFactory getInstance() {
+    return instance;
+  }
+
+  @Override
+  public <T> LoadBalancer<T> newLoadBalancer(String serviceName, TransportManager<T> tm) {
+    return new RoundRobinLoadBalancer<T>(tm);
+  }
+
+  private static class RoundRobinLoadBalancer<T> extends LoadBalancer<T> {
+    private static final Status SHUTDOWN_STATUS =
+        Status.UNAVAILABLE.augmentDescription("RoundRobinLoadBalancer has shut down");
+
+    private final Object lock = new Object();
+
+    @GuardedBy("lock")
+    private RoundRobinServerList<T> addresses;
+    @GuardedBy("lock")
+    private InterimTransport<T> interimTransport;
+    @GuardedBy("lock")
+    private Status nameResolutionError;
+    @GuardedBy("lock")
+    private boolean closed;
+
+    private final TransportManager<T> tm;
+
+    private RoundRobinLoadBalancer(TransportManager<T> tm) {
+      this.tm = tm;
+    }
+
+    @Override
+    public T pickTransport(Attributes affinity) {
+      synchronized (lock) {
+        if (closed) {
+          return tm.createFailingTransport(SHUTDOWN_STATUS);
+        }
+        if (addresses == null) {
+          if (nameResolutionError != null) {
+            return tm.createFailingTransport(nameResolutionError);
+          }
+          if (interimTransport == null) {
+            interimTransport = tm.createInterimTransport();
+          }
+          return interimTransport.transport();
+        }
+      }
+      return addresses.getTransportForNextServer();
+    }
+
+    @Override
+    public void handleResolvedAddresses(
+        List<? extends List<ResolvedServerInfo>> updatedServers, Attributes config) {
+      InterimTransport<T> savedInterimTransport;
+      synchronized (lock) {
+        if (closed) {
+          return;
+        }
+        RoundRobinServerList.Builder<T> listBuilder = new RoundRobinServerList.Builder<T>(tm);
+        for (List<ResolvedServerInfo> servers : updatedServers) {
+          if (servers.isEmpty()) {
+            continue;
+          }
+
+          final List<SocketAddress> socketAddresses = new ArrayList<SocketAddress>(servers.size());
+          for (ResolvedServerInfo server : servers) {
+            socketAddresses.add(server.getAddress());
+          }
+          listBuilder.addAll(socketAddresses);
+        }
+        RoundRobinServerList<T> newRoundRobin = listBuilder.build();
+        if (newRoundRobin.equals(addresses)) {
```

Comment:
I would avoid defining the nontrivial equality for RoundRobinServerList, because it has state.
We don't have to handle the "same addresses received" case here. It should be NameResolver's responsibility to not give spurious updates. There is a case where addresses are not changed while the config is, but that should be infrequent enough so that resetting the round-robin pointer wouldn't be an issue.